classes/ActionScheduler_wpCommentCleaner.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/**
+ * Class ActionScheduler_WPCommentCleaner
+ *
+ * @since 3.0.0
+ */
+class ActionScheduler_WPCommentCleaner {
+
+	/**
+	 * Post migration hook used to cleanup the WP comment table.
+	 *
+	 * @var string
+	 */
+	protected static $cleanup_hook = 'action_scheduler/cleanup_wp_comment_logs';
+
+	/**
+	 * An instance of the ActionScheduler_wpCommentLogger class to interact with the comments table.
+	 *
+	 * This instance should only be used as an interface. It should not be initialized.
+	 *
+	 * @var ActionScheduler_wpCommentLogger
+	 */
+	protected static $wp_comment_logger = null;
+
+	/**
+	 * Initialize the class and attach callbacks.
+	 */
+	public static function init() {
+		if ( empty( self::$wp_comment_logger ) ) {
+			self::$wp_comment_logger = new ActionScheduler_wpCommentLogger();
+		}
+
+		add_action( self::$cleanup_hook, array( __CLASS__, 'delete_all_action_comments' ) );
+
+		// While there are orphaned logs left in the comments table, we need to attach the callbacks which filter comment counts.
+		add_action( 'pre_get_comments', array( self::$wp_comment_logger, 'filter_comment_queries' ), 10, 1 );
+		add_action( 'wp_count_comments', array( self::$wp_comment_logger, 'filter_comment_count' ), 20, 2 ); // run after WC_Comments::wp_count_comments() to make sure we exclude order notes and action logs
+		add_action( 'comment_feed_where', array( self::$wp_comment_logger, 'filter_comment_feed' ), 10, 2 );
+
+		// Action Scheduler may be displayed as a Tools screen or WooCommerce > Status administration screen
+		add_action( 'load-tools_page_action-scheduler', array( __CLASS__, 'print_admin_notice' ) );
+		add_action( 'load-woocommerce_page_wc-status', array( __CLASS__, 'print_admin_notice' ) );
+	}
+
+	/**
+	 * Determines if there are log entries in the wp comments table.
+	 *
+	 * @return boolean Whether there are scheduled action comments in the comments table.
+	 */
+	public static function has_logs() {
+		return (bool) get_comments( array( 'type' => ActionScheduler_wpCommentLogger::TYPE, 'number' => 1, 'fields' => 'ids' ) );
```

Comment:
Are `get_comments()` queries cached?

If not, we need to add our own caching layer here (a transient with a short expiration is fine for that) because right now, `ActionScheduler_WPCommentCleaner::has_logs()` is being called on every request and it could be quite slow on large sites.

Reply:
They are cached using the `wp_cache_()` APIs so it runs once per request. So I think it's a good idea to add a transient.

Reply:
I've pushed 98eb0cf to cache whether there are logs in the comments table. 

When I was considering an expiration for it I realised that because we are no longer writing AS logs to the comments table, it's safe to assume the first value we find will be correct until the logs are cleared out either by the scheduled action or by someone running a database query manually.

If there are no AS comments left, that's not going to change so do we need to put an expiry on it? It should remain no forever.

If it's yes, then it's going to remain yes until they are deleted. 

In that case, should this be an option rather than a transient?

If we use an option, we can also avoid having this option/transient live forever - we can delete the option once we're done.

Reply:
I've pushed b43b9b2 to use an option rather than a transient. Let me know what your opinion is. 

> If we use an option, we can also avoid having this option/transient live forever - we can delete the option once we're done.

With the changes in b43b9b2, the option will still live forever. To avoid that we would need to:

1. Run the `get_comments()` call once on migration complete. At this point set the option.
2. After that point, if there's an option value stored we assume there are comment logs, if there isn't an option stored, we assume there aren't comment logs.

I'm not sure if it's worth going down that route. Let me know what you think though. 
It would basically be equivalent to checking if there's a clean up scheduled action scheduled; they would both be scheduled/set and run/deleted at the same time.

Reply:
> I'm not sure if it's worth going down that route.

I think it's worth it. It's a one off thing, so worth removing all trace once it's done.

> It would basically be equivalent to checking if there's a clean up scheduled action scheduled

Yeah, that's true, but the data does serve two different purposes - one is to schedule a deletion, one is (very performant) a flag to indicate some state exists.

> When I was considering an expiration for it

I love everything that follows there haha.

Reply:
hahah ok. I've squashed all those commits and moved to the option set once on migration complete approach.